### PR TITLE
fix infinite scroll issue

### LINKED ISSUE #114: infinite scroll doesn't display all results
### Steps to Reproduce

Visit https://ocwnext.odl.mit.edu/search/?q=%22quantum+mechanics%22 and continuously scroll to the bottom.  Open an inspector and compare the last `article` element's `aria-labelledby` property to the total amonut of results listed at the top of the page

### Expected Behavior

The page renders all of the results that are supposed to be returned by the search terms

### Actual Behavior

An inconsistent amount of search results are actually rendered

### Screenshot or Screencast

image

### PR DESCRIPTION
**Pull request by aliceriot**
#### Pre-Flight checklist

- [x] Screenshots and design review for any changes that affect layout or styling
  - [x] Desktop screenshots
  - [x] Mobile width screenshots
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?

closes #114

#### What's this PR do?

This fixes the infinite scroll issue we were seeing by applying two fixes:

- ensuring that the state update for new search result data is always applied as an update to the most up-to-date data
- fixing the logic for the `loaded` prop passed to `useCourseSearch` so that only one request may be made at a time

This ensures that we display all the search results and that they are displayed in the right order. It also ensures we're not putting unnecessary, bursty load on the backend by firing off like 15 search requests at the same time 😃 

#### How should this be manually tested?

It seems the easiest way to test this if you don't have a fully set-up open discussions index locally is to point your ocw-hugo-themes repo at the RC search.

You can do that by setting:

```
SEARCH_API_URL=$DISCUSSIONS_RC/api/v0/search/
```

Then you'll face a CORS issue. You can either use a browser extension to turn off CORS temporarily, or you can pass a command-line flag to Chrome to open it without CORS. Probably not a good idea to then do anything else in this window at all!

The command to do that looks like:

```
/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --disable-web-security --user-data-dir=~/chromeTemp/
```

on macOS, and on Linux you can probably just do something like

```
google-chrome --disable-web-security --user-data-dir=~/chromeTemp/
```

I think this would probably work with Chromium too.

#### Screenshots (if appropriate)

Screen Shot 2021-06-17 at 9 57 31 AM
Screen Shot 2021-06-17 at 9 58 18 AM

### CODE REVIEW

File: www/assets/js/components/SearchPage.js [hunk 1]
```diff
@@ -76,21 +76,23 @@ export default function SearchPage() {
 
       setSearchFacets(new Map(Object.entries(newResults.aggregations ?? {})))
 
-      setSearchResults(
+      setSearchResults(results =>
```
Comment (noisecapella) [inline www/assets/js/components/SearchPage.js:79]: That's interesting, I wish they would highlight the need for this in the docs for handling previous state

**Review by noisecapella — CHANGES_REQUESTED** [review]
If I make a change to a facet or to the sort dropdown it shows a loading indicator indefinitely

**Comment by aliceriot** [comment]
aha ok well you can tell I tested thoroughly :P

I'll figure out what's up and ping you

**Comment by aliceriot** [comment]
ok I think that should be fixed - I forgot that we reset `completedInitialLoad` to `false` after clearing the search because of a facet change.

File: www/assets/js/components/SearchPage.js (hunk 1 above)
Comment (aliceriot) [reply www/assets/js/components/SearchPage.js:79]: yeah...they discuss it in the docs but don't really clarify how helpful it is for avoiding race conditions for these kinds of things. It also means if you're passing the state update function in to a `useCallback` function you can avoid also passing the current value as a dependency, which I think would in a lot of cases drastically cut down on the number of unnecessary re-allocations of that function (which will increase memory usage and pressure on the GC and whatnot).

**Review by noisecapella — APPROVED** [review]
Looks good now :+1:

**Pull request merged**